### DISCUSSION TITLE
Retrieval of labels (street names, city names etc.)?

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/mapsforge#1296] [Q&A] [answered]
Is there a way to retrieve labels like street names, city names etc. with their coordinates from an existing map file?
If not, could you give me a pointer where to start to extend the library to add such a function?

I want to build a POI database with this info to allow search for street names, city names etc., so after installing a map file there would be some sort of map file scan and database build, after which such a search function could be offered by using the data stored in the database.

**Comment by devemux86** [comment, marked as answer]
> Is there a way to retrieve labels like street names, city names etc. with their coordinates from an existing map file?

Mapsforge map format is designed for fast map rendering (not search).
The closest is reverse geocoding, see the [ReverseGeocodeViewer](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/ReverseGeocodeViewer.java) sample.

> I want to build a POI database with this info to allow search

There is the [Mapsforge POI API](https://github.com/mapsforge/mapsforge/blob/master/docs/POI.md), see the [PoiSearchViewer](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/PoiSearchViewer.java) sample.

_But it does not work with file streams on Android scoped storage._
Pull requests are always welcome :slightly_smiling_face: 

Poi files are available in [Mapsforge server](https://download.mapsforge.org/) and [OpenAndroMaps](https://www.openandromaps.org/en).

**Reply by moving-bits** [reply]
Thanks for your reply.

> Mapsforge map format is designed for fast map rendering (not search).

I'd read that somewhere while looking for possible solutions, therefore my idea to do a manual scan for existing labels once, and put that info into a database. All searches will then be run against the database. (Somehow similar to what is done with POI.)

> The closest is reverse geocoding, see the [ReverseGeocodeViewer](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/ReverseGeocodeViewer.java) sample.

Which data does it get its results from - a POI database or the labels included in the map file?

I've played around a bit with it, but it does not seem to deliver the information I'm looking for. E. g.: I've zoomed the map to display a couple of streets. Some streets have a label (street name), but even then the result being displayed stays empty (at least in the "ways" section).

> > I want to build a POI database with this info to allow search
> 
> There is the [Mapsforge POI API](https://github.com/mapsforge/mapsforge/blob/master/docs/POI.md), see the [PoiSearchViewer](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/src/main/java/org/mapsforge/samples/android/PoiSearchViewer.java) sample.

This was my first approach. However it contains exactly that, POI. When I'm looking for specific streets, then I can find this street only, when there's a POI in this street. Streets without POI cannot be found by this method.

That's why I started looking at the map file. It should be all in there, street names, city names etc., and also having info for label type (e. g. "street", "city") and coordinates. I just need a find a way to extract that info... Currently I do not know enough about the map file format and how to access its contents. I guess there's something like a hierarchical structure. Then I would need to traverse it, look for labels and collect the data along the way. Do you have a pointer for me into the code how to traverse the data?

> _However it does not work with file streams on Android scoped storage._ Pull requests are always welcome 🙂

I'm certainly willing to provide a PR, just need to build some to create a PR from first :-)

**Reply by devemux86** [reply]
> Which data does it get its results from

Please study the sample, it just asks map tiles for their data.
Note that map files store only data useful for map rendering.

> but it does not seem to deliver the information I'm looking for.

The sample filters nodes and closed ways.

> put that info into a database.

POI API reads OSM data and stores it in SQLite.

> However it contains exactly that, POI.

It can contain anything you store.
e.g. calculate externally / store road centroids.

---

As already mentioned:
_Mapsforge map format is designed for fast map tile rendering (not any kind of search)._
It is better to use the appropriate formats for each work.

**Reply by moving-bits** [reply]
I'm beginning to understand, thanks for your patience and for the pointers. Filtering out closed ways was what confused me. If I look into category `ways` of the `MapReadResult` and remove the filter for `isClosedWay`, then I get much more info, including streets and administrative levels.

My approach would now be:
- create a background service to scan a selected map file (as this can be a very time consuming function)
- Scanning does a `readLabels()` for the whole area covered by the selected map file.
- Build a database from the results, storing coordinates, label and type of label.
- As a first approach only labels having a non-empty attributes of either `highway` or `admin_level` would be considered.
- Searching would then be executed against the database.

I'll also look into the POI API to see whether it can help with some of the steps above.

My objective is to create a client-side extension to our geocaching app c:geo to enable users to search for streets/city names etc. using the map file(s) users have downloaded on their system (which may vary widely in both covered areas and sources they have downloaded their files from). It's not my intention to provide extra POI files for download.

**Reply by devemux86** [reply]
Reverse geocoding works with map tiles, so it depends on zoom level.
As Mapsforge map files may contain different data per [zoom interval](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md).

**Reply by moving-bits** [reply]
Do higher zoom levels include all labels of lower zoom levels per definition (in context of `readLabels()`), or is it completely independent?

**Reply by devemux86** [reply]
The higher zoom interval should contain all rendering data.
For more details please see the [map-writer documentation](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md).
